tests/python_package_test/test_dask.py

```diff
@@ -826,3 +826,31 @@ def test_dask_classes_and_sklearn_equivalents_have_identical_constructors_except
     assert dask_spec.defaults[:-1] == sklearn_spec.defaults
     assert dask_spec.args[-1] == 'client'
     assert dask_spec.defaults[-1] is None
+
+
+@pytest.mark.parametrize(
+    "methods",
+    [
+        (lgb.DaskLGBMClassifier.fit, lgb.LGBMClassifier.fit),
+        (lgb.DaskLGBMClassifier.predict, lgb.LGBMClassifier.predict),
+        (lgb.DaskLGBMRegressor.fit, lgb.LGBMRegressor.fit),
+        (lgb.DaskLGBMRegressor.predict, lgb.LGBMRegressor.predict),
+        (lgb.DaskLGBMRanker.fit, lgb.LGBMRanker.fit),
+        (lgb.DaskLGBMRanker.predict, lgb.LGBMRanker.predict)
+    ]
+)
+def test_dask_methods_and_sklearn_equivalents_have_similar_signatures(methods):
+    dask_spec = inspect.getfullargspec(methods[0])
+    sklearn_spec = inspect.getfullargspec(methods[1])
+    dask_params = inspect.signature(methods[0]).parameters
+    sklearn_params = inspect.signature(methods[1]).parameters
+    assert dask_spec.args == sklearn_spec.args[:len(dask_spec.args)]
+    assert dask_spec.varargs == sklearn_spec.varargs
+    if sklearn_spec.varkw:
+        assert dask_spec.varkw == sklearn_spec.varkw[:len(dask_spec.varkw)]
+    assert dask_spec.kwonlyargs == sklearn_spec.kwonlyargs
+    assert dask_spec.kwonlydefaults == sklearn_spec.kwonlydefaults
+    for key in dask_params:
+        if key in sklearn_params:
```

Comment:
Please remove this if statement. If the Dask side adds an argument that is not in the equivalent scikit-learn side, we want this test to fail.

Reply:
Yeah I was having a doubt on this one. Isn't the test already failing for the `fit()` method? Because the dask version has `**kwargs` pass-through , but the sklearn version does not. When I run the code with the if statement, it gives the error:

`Keyerror: 'kwargs'` 
Which makes sense as the `fit` methods in sklearn's estimators have no variable keyword arguments.

If I get any `Keyerror` in the mentioned line, it means that the particular dask param has no sklearn coumterpart.
I propose to just catch the `Keyerror` and the `AssertionError` and display different error messages for them, for example

```suggestion
        try:
            assert dask_params[key].defaults == sklearn_params[key].defaults
        except AssertionError:
            print(error_msg1)
        except Keyerror:
            print(f"param {key} in dask's method has no sklearn equivalent")
```

Reply:
Yeah sure. Its already failing the test though right? Because the `fit()` method of dask has `**kwargs` passthrough, but the sklearn method does not. When I run the assertion, it gives the following error for only the `fit` methods:

`Keyerror: 'kwargs'`

Which makes sense considering that sklearn `fit()` has no variable keyword arguments

Reply:
I tried to use an additional assertion to catch this , something like 
`error_msg2 = f"param '{key}' in dask's method has no sklearn equivalent"`
`assert key in sklearn_params is True, error_msg2`

Reply:
Are you sure? I just ran the following:

```python
import lightgbm as lgb
import inspect

dask_spec = inspect.getfullargspec(lgb.DaskLGBMClassifier.fit)
print(dask_spec.args)
```

and `kwargs` was not included in the list

> ['self', 'X', 'y', 'sample_weight']

Reply:
I meant the `inspect.signature` method, 
```
import lightgbm as lgb
import inspect

dask_params = inspect.signature(lgb.LGBMClassifier.predict).parameters
print(dask_params)
```

The code above gives the following output
```
OrderedDict([('self', <Parameter "self">), ('X', <Parameter "X">), ('raw_score', <Parameter "raw_score=False">), ('start_iteration', <Parameter "start_iteration=0">), ('num_iteration', <Parameter "num_iteration=None">), ('pred_leaf', <Parameter "pred_leaf=False">), ('pred_contrib', <Parameter "pred_contrib=False">), ('kwargs', <Parameter "**kwargs">)])
```

`kwargs` is included

Reply:
Oh yeah you mentioned it before, I'll change the loop to loop over `dask_spec.args` instead of `dask_params`

Reply:
ah I see! `inspect` has several functions with names that sound like they do the same thing, so it can get confusing. If you loop over `.args` like I suggested in https://github.com/microsoft/LightGBM/pull/3911#discussion_r571111631, this check won't be necessary and we can rely on getting a KeyError if there are arguments in the Dask inreface not found in the sklearn equivalent

Reply:
Yes exactly, I realised that

Reply:
We only needed to check the positional arguments for their defaults, as we already had checked for keyword only defaults earlier